subs/pantry/src/Pantry/Types.hs

```diff
@@ -1638,6 +1638,7 @@ data HpackExecutable
 -- @since 0.1.0.0
 data WantedCompiler
   = WCGhc !Version
+  | WCGhcGit !Text !Text
```

Comment:
This is honestly the most controversial thing here IMO. I'm not sure if snapshots should have the ability to specify arbitrary GHC commits like this, though I understand the motivation. I'd also recommend replacing `Text` with more type-safe values. In particular, `data Flavour = FlavourQuick` would seem reasonable.

Reply:
Controversial in the sense of "this isn't obviously the right place for it," nothing more. It might make more sense to treat these Git-built GHCs like any other GHC instead of treating it as a completely separate compiler. For example, if I build `ghc-8.6.4.23` from some branch, and a Stackage snapshots wants `ghc-8.6.4`, should the Git-built one work for that or not? As you've set it up here, we're treating these things as completely separate compilers.

See #4637 for a related issue which would be impacted by the decision here

Reply:
Ok I see. There are several issues:

1) The Version type isn't adapted and has deep roots into Cabal etc.: i've thought about unifying ``WCGhc`` and ``WCGhcGit`` by using something like [versionTags](https://www.stackage.org/haddock/lts-13.16/base-4.12.0.0/Data-Version.html#v:versionTags) to indicate the commit and the flavour, but since Cabal 2.0 [we don't have these tags anymore](https://www.stackage.org/haddock/lts-13.16/Cabal-2.4.1.0/Distribution-Make.html#t:Version).

2) Retrieving a meaningful GHC version is hard: only release branches (e.g. "ghc-8.6" branch) ensure that the ABI stays the same and it's just discipline by GHC HQ (we can imagine a custom GHC repo where it is not true). GHCs built from other branches such as master may return meaningless versions (e.g. ghc from "master" returns version "8.9.20190410" where the date is the date when ``configure`` was run...) and the ABI may be broken anytime.

So I don't think it is worth the effort to try to detect when there is a compatibility between the Git built compiler and the Stackage snapshot. Except maybe when there is a perfect match between Version and CommitID+Flavour.

---

The other way around is simpler (from Version to Commit/Flavour) but not trivial either. To implement #4637 we can imagine retrieving the Git tags from the ``compiler-repository``: e.g. to build 8.6.4 from source, we infer that it correspond to tag "ghc-8.6.4-release" (i.e. commit aac18e9..) built with "perf" flavour.

It's not always true though: e.g. 8.6.4 bindist on Windows has been released by mistake while built with the "quick" flavour; we would need to handle this quirk to get the same behavior between bindist/source versions :/ or FP Complete might distribute its own bindists ensured to be built with the "perf" flavour, but this is another can of worms...

So for now I think keeping both compilers separate is better.

Reply:
OK, that makes sense. To be clear, I definitely didn't have what I thought was a better solution to the problem. Let's stick with this for now. Thank you for clarifying :)